### PR TITLE
Remove unnecessary differences from upstream `AntClassLoader`

### PR DESCRIPTION
**Pull request by basil**
This is the first of a series of changes to unfork `AntClassLoader`, a class which we copied from Ant a long time ago (latest [upstream](https://github.com/apache/ant/blob/cc9d93f49fbd6730e057ec0ad6e2e382f8e3c331/src/main/org/apache/tools/ant/AntClassLoader.java)) and has since become a bit of a maintenance burden. This PR does two things:

- Undo non-substantive refactorings that we've done to the fork that aren't present upstream. These didn't buy us much in terms of functionality, but they increased our maintenance burden by increasing the difference between our fork and upstream. For example, I undid the move from `VectorSet` to `ArrayList` and the change that factored out some duplicate code into a `storeAndConvert` method.
- Apply non-substantive refactorings that upstream has done but which our fork doesn't have. These don't buy us much in terms of functionality, but they decrease our maintenance burden by minimizing the difference between our fork and upstream. This includes Java generic support, implementing `Closeable`, adding the `final` keyword to everything, using the Stream API in a few places, and adding some new Javadocs.

With the above taken care of, it's possible to see clearly what Jenkins-specific changes remain:

- Some `super` calls were added to fix [JENKINS-21579](https://issues.jenkins.io/browse/JENKINS-21579)/[JENKINS-22310](https://issues.jenkins.io/browse/JENKINS-22310).
- A `getUrl` method was factored out for use in `AntWithFindResourceClassLoader`.
- `findResources(String)` was made public for use in `ClassLoaderReflectionToolkit`.
- Missing class telemetry support was added.

Here are the remaining upstream changes that were _not_ applied to this PR because I judged them to be too risky:

- commit 593aff2d2 bz-62952 Make `AntClassLoader` multi-release JAR aware when it deals with `java.util.jar.JarFile`
- commit 189ad7e59 bz-62764 Add a debug log message when we skip invalid path elements from the `AntClassLoader`'s classpath
- commit 008f1c8be Embrace `StringUtils#getStackTrace`
- commit d37df73a8 Register `AntClassLoader` as parallel capabale
- commit 13c01b296 Improve skipping of non-zips

I'll probably backport these onto our copy once we start a new LTS cycle.

### Proposed changelog entries

N/A

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by basil** [comment]
I realize this change is impossible to review. Might it be possible to get a vote of confidence that I generally know what I'm doing, and if I don't then I'm responsible enough to deal with the consequences? We're at the beginning of a new LTS cycle, so it seems like a good time to kick off this unforking mini-project.

**Review by timja — APPROVED** [review]
looks fine, mostly fairly cosmetic changes

**Review by res0nance — APPROVED** [review]

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Pull request merged**
